test/jdk/javax/net/ssl/SSLParameters/NamedGroups.java

```diff
@@ -0,0 +1,147 @@
+ * @library /javax/net/ssl/templates
+ * @run main/othervm NamedGroups
+ */
+
+import javax.net.ssl.SSLParameters;
+import javax.net.ssl.SSLServerSocket;
+import javax.net.ssl.SSLSocket;
+import java.security.Security;
+
+public class NamedGroups extends SSLSocketTemplate {
+    private final String[] serverNamedGroups;
+    private final String[] clientNamedGroups;
+    private final boolean exceptionExpected;
+
+    public NamedGroups(String[] serverNamedGroups,
+                            String[] clientNamedGroups,
+                            boolean exceptionExpected) {
+        this.serverNamedGroups = serverNamedGroups;
+        this.clientNamedGroups = clientNamedGroups;
+        this.exceptionExpected = exceptionExpected;
+    }
+
+    @Override
+    protected void configureServerSocket(SSLServerSocket sslServerSocket) {
+        SSLParameters sslParameters = sslServerSocket.getSSLParameters();
+        sslParameters.setNamedGroups(serverNamedGroups);
+        sslServerSocket.setSSLParameters(sslParameters);
+    }
+
+    @Override
+    protected void configureClientSocket(SSLSocket socket) {
+        SSLParameters sslParameters = socket.getSSLParameters();
+        sslParameters.setNamedGroups(clientNamedGroups);
+        socket.setSSLParameters(sslParameters);
+    }
+
+    @Override
+    protected void runServerApplication(SSLSocket socket) {
+        try {
+            super.runServerApplication(socket);
+        } catch (Exception ex) {
+            // Just ignore, let the client handle the failure information.
+        }
+    }
+
+    @Override
+    protected void runClientApplication(SSLSocket sslSocket) throws Exception {
+        try {
+            super.runClientApplication(sslSocket);
+        } catch (Exception ex) {
+            if (!exceptionExpected) {
+                throw ex;
+            } else {  // Otherwise, swallow the exception and return.
+                return;
+            }
+        }
+
+        if (exceptionExpected) {
+            throw new RuntimeException("Unexpected success!");
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        Security.setProperty("jdk.tls.disabledAlgorithms", "");
+
+        runTest(new String[] {
+                        "x25519",
+                        "secp256r1"
+                },
+                new String[] {
+                        "x25519",
+                        "secp256r1"
+                },
+                false);
+        runTest(new String[] {
+                        "secp256r1"
+                },
+                new String[] {
+                        "secp256r1"
+                },
+                false);
+        runTest(null,
+                new String[] {
+                        "secp256r1"
+                },
+                false);
+        runTest(new String[] {
+                        "secp256r1"
+                },
+                null,
+                false);
+        runTest(new String[0],
```

Comment:
I think this case will throw IAE when `SSLParameter.setNamedGroups()` is called right? If so, I think you can delete this test (and the one one on line 123) since it is already covered in the NamedGroupsSpec test. Also, this test is coded such that the expected exception is thrown when the TLS handshake is being made in `runClientApplication()`, which is not the case here.

Reply:
It is allowed to set empty named groups with `SSLParameter.setNamedGroups()`, which means (see spec at `getNamedGroups`) that 

```
     * If the returned array is empty (zero-length), then the named group
     * negotiation mechanism is turned off for SSL/TLS/DTLS protocols, and
     * the connections may not be able to be established if the negotiation
     * mechanism is required by a certain SSL/TLS/DTLS protocol.
```

The TLS connection will fail, as expected, although for this case.